### PR TITLE
[clang][NFC] Add CWG713 test; add example from CWG1584 test

### PR DESCRIPTION
**Pull request by MitalAshok**
https://cplusplus.github.io/CWG/issues/713.html
https://cplusplus.github.io/CWG/issues/1584.html

### CODE REVIEW

**Comment by MitalAshok** [comment]
@Endilll The problem with the CWG713 test in Clang 18 was with parsing the type trait (#81298), they weren't actually considered `const`: https://godbolt.org/z/snde76brE

Also, I would think since the resolution is purely editorial (it just changes notes, nothing normative), it would be considered implemented in any clang version?

**Comment by Endilll** [comment]
> Also, I would think since the resolution is purely editorial (it just changes notes, nothing normative), it would be considered implemented in any clang version?

The question "availability" answers is "What is the first Clang version that exhibits the correct behavior?", where the correct behavior is approximated by the test in our test suite. Because Clang 18 does not pass your CWG713 test, we can't claim that it implements this CWG issue as we understand it.

File: clang/test/CXX/drs/cwg7xx.cpp [hunk 1]
```diff
@@ -71,6 +76,17 @@ namespace cwg712 { // cwg712: partial
 #endif
 }
 
+namespace cwg713 { // cwg713: yes
```
Comment (Endilll) [inline clang/test/CXX/drs/cwg7xx.cpp:79]: Please check how just released Clang 19.1-rc1 handles this test, and choose between 19 and 20 accordingly.

**Review by Endilll — CHANGES_REQUESTED** [review]

**Comment by MitalAshok** [comment]
New test compiles down to clang 3.0 (except warnings for https://cplusplus.github.io/CWG/issues/547.html) but do not compile before that because of the lack of `__is_const`/`__is_volatile` type traits.

Function types with cv- qualifiers were added for Clang 2.5 in 22c40fa28551f0da487f399a2fe0d3d25edf3edc, where they already used a different field to hold the qualifiers (thus the function types were not considered `const`/`volatile`/`__restrict` themselves), so it should be okay to claim this was implemented in Clang 2.5?

**Comment by Endilll** [comment]
> New test compiles down to clang 3.0 (except warnings for https://cplusplus.github.io/CWG/issues/547.html) but do not compile before that because of the lack of `__is_const`/`__is_volatile` type traits.
> 
> Function types with cv- qualifiers were added for Clang 2.5 in [22c40fa](https://github.com/llvm/llvm-project/commit/22c40fa28551f0da487f399a2fe0d3d25edf3edc), where they already used a different field to hold the qualifiers (thus the function types were not considered `const`/`volatile`/`__restrict` themselves), so it should be okay to claim this was implemented in Clang 2.5?

Thank you for the analysis! Typically you don't need to dig that deep. Based on https://godbolt.org/z/n6oPdMT91, I think `3.1` is the correct availability here, but I can be persuaded for `3.0`.

Clang 2.7 was the first one that by default provided C++ support without relying on GCC frontend, so I don't think it's meaningful to go further than that. (`Clang 2.6 (assertions)` on Compiler Explorer is patched to use internal C++ support, you can see for yourself how broken it was.)

**Comment by MitalAshok** [comment]
> Based on https://godbolt.org/z/n6oPdMT91, I think `3.1` is the correct availability here, but I can be persuaded for `3.0`.

The warnings are because of how the test is written, the tested behaviour ("function types are never const or volatile") is still true in Clang 3.0

**Review by Endilll — APPROVED** [review]

**Pull request merged**